engine/src/board.rs

```diff
@@ -248,55 +275,71 @@ impl Board {
         false
     }
 
-    fn walk_board(
-        &self,
-        position: Position,
-        excluded_position: Position,
-        mut visited: HashSet<Position>,
-    ) -> HashSet<Position> {
-        self.walk_board_inner(position, excluded_position, &mut visited);
-        visited
-    }
-
-    fn walk_board_inner(
-        &self,
-        position: Position,
-        excluded_position: Position,
-        visited: &mut HashSet<Position>,
-    ) {
-        if visited.contains(&position) {
-            return;
-        }
-        visited.insert(position);
-        for pos in self.positions_taken_around_iter(position) {
-            if pos != excluded_position && !visited.contains(&pos) {
-                self.walk_board_inner(pos, excluded_position, visited);
-            }
+    pub fn update_pinned(&mut self) {
+        for pinned_info in self.calculate_pinned().iter() {
+            self.pinned[self.piece_to_offset(pinned_info.piece)] = pinned_info.pinned
         }
     }
 
-    pub fn pinned(&self, position: Position) -> bool {
-        // pieces on top of the hive cannot be pinned (just gated)
-        if self.level(position) > 1 {
-            return false;
-        }
-        // if there's only one neighbor the piece isn't pinned
-        let all_neighbor_positions = self
-            .positions_taken_around_iter(position)
+    pub fn calculate_pinned(&self) -> Vec<PinnedInfo> {
+        // make sure to get only top pieces in this
```

Comment:
Comment says top pieces, but shouldn't it be bottom pieces? 

As top pieces above base level can't be pinned, and this way by doing top_piece any b/m on top of a pinned piece would get flagged as pinned I think